### PR TITLE
Core/Pools: fix pool memory usage

### PR DESCRIPTION
**Pull request by Rochet2**
**Changes proposed:**

-  Change std::vector to std::unordered_map to save memory due to only sparse keys used from the container
- This should slightly speed up startup
- This should slightly speed up saving weekly and daily quests
- This should reduce ram usage quite significantly (10% saved after startup)
- Pool loading in startup can look slower in the log messages due to unordered map find and space allocated in pool loading rather than initialization
- This may slightly slow the pool system since vector random access changed to unordered map find
- Remove max_pool_id since it is unused
- Add some startup checks that log pool IDs that are not in pool_template (before only < maxpoolid was checked). **The ones that do not exist in pool template are skipped**
- **Assumption**: All pool_id are supposed to be in pool_template
- One of the alternative fixes to changing code was to fix the DB data instead of core code. Read related issue comments: https://github.com/TrinityCore/TrinityCore/issues/18661#issuecomment-269690058

If the DB approach is preferred, then some guidance on how the data transformation could be done would be nice.
For example should all existing rows be deleted and new data be inserted, or should we just use SET statements to set the pool ids to new ones? What about custom data?

One idea could be to make a script that takes all custom data, puts it to the side, reorders TC data by pool_id, appends custom data on top of that (with reorder).

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [x] master

**Issues addressed:**
Closes https://github.com/TrinityCore/TrinityCore/issues/18661

**Tests performed:** (Does it build, tested in-game, etc.)
- Compiles.
- Logged ingame, looted gameobjects and herbs and waited for them to respawn. (silverpine forest, nagrand, hellfire peninsula)
- Checked that startup times after change were slightly faster. (this can vary significantly since there can be even 30 seconds of difference between two startups of same server) Here is a before and after of the pool loading: https://i.imgur.com/ZoSSodh.png and it disregards initialization time of pools for before.
- Checked that startup errors popped up about entries that are not in pool_template but are used in other pool tables.
- Checked that RAM consumed was around 1833 MB before and around 1649 MB after change. (multiple startups of the server)

**Known issues and TODO list:** (add/remove lines as needed)
- [x] Choose between DB and core edits.

### CODE REVIEW

**Comment by jackpoz** [comment]
is the increased ram usage caused by those high entry values in the database ? could the ram usage issue be fixed by compacting the ids in the db ?

**Comment by Shauren** [comment]
Yes but tbh using a plain vector for this is a terrible idea when ids don't have to be contiguous

**Comment by Takenbacon** [comment]
@Shauren It's always a trade off with performance vs ram usage, how much slower is pooling now with having map lookups?

File: src/server/game/Pools/PoolMgr.h [hunk 1]
```diff
@@ -141,12 +141,11 @@ class TC_GAME_API PoolMgr
...
         void SpawnPool(uint32 pool_id, uint32 db_guid_or_pool_id);
 
-        uint32 max_pool_id;
-        typedef std::vector<PoolTemplateData>       PoolTemplateDataMap;
-        typedef std::vector<PoolGroup<Creature> >   PoolGroupCreatureMap;
-        typedef std::vector<PoolGroup<GameObject> > PoolGroupGameObjectMap;
-        typedef std::vector<PoolGroup<Pool> >       PoolGroupPoolMap;
-        typedef std::vector<PoolGroup<Quest> >      PoolGroupQuestMap;
+        typedef std::unordered_map<uint32, PoolTemplateData>       PoolTemplateDataMap;
+        typedef std::unordered_map<uint32, PoolGroup<Creature> >   PoolGroupCreatureMap;
```
Comment (ariel-) [inline src/server/game/Pools/PoolMgr.h]: Since this is new code and we support c++11 there is no need to leave a space between template brackets

**Comment by Rochet2** [comment]
@Takenbacon Vector and map access time is the same when doing 1-10 accesses in succession.
100000000 accesses takes 0ms with vector and 1 second with map, so over the course of maybe 20 days you could lose one second if the pool system maps are actively called (like 50 times every second).


I did not make any exhaustive performance analysis on the pooling system, but based on this I would say its not really slower.

**Comment by Shauren** [comment]
The point is, you don't iterate over pool vectors (which is what would give it the largest advantage)

**Comment by jackpoz** [comment]
so let's merge it ?

**Comment by jackpoz** [comment]
@Rochet2 please don't go on vacation this Xmas in case some issues arise :)

**Pull request merged**